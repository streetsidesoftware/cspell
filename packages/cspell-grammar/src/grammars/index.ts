export * as Markdown from './markdown.js';
export * as TypeScript from './typescript.js';
