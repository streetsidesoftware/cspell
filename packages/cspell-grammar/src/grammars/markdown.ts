/* eslint-disable no-useless-escape */
import type { GrammarDef, Repository } from '..';

const repository: Repository = {
    frontMatter: { patterns: [] },
    block: { patterns: ['#raw'] },

    fenced_code_block: {
        patterns: ['#fenced_code_block_unknown'],
    },
    fenced_code_block_unknown: {
        name: 'markup.fenced_code.block.markdown',
        begin: /(^)(\s*)(`{3,}|~{3,})\s*(?=([^`~]*)?$)/,
        beginCaptures: {
            '3': 'punctuation.definition.markdown',
            '4': 'fenced_code.block.language',
        },
        // eslint-disable-next-line no-useless-backreference
        end: /(^|\G)(\2|\s{0,3})(\3)\s*$/,
        endCaptures: {
            '3': 'punctuation.definition.markdown',
        },
    },
    raw: {
        name: 'markup.inline.raw.string.markdown',
        captures: 'punctuation.definition.raw.markdown',
        begin: /`+/,
        end: '\\1',
    },
};

export const grammar: GrammarDef = {
    name: 'Markdown',
    scopeName: 'text.html.markdown',
    patterns: ['#frontMatter', '#block'],
    repository,
};
