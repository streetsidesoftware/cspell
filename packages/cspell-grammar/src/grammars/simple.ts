import type { GrammarDef, Repository } from '..';

const repository: Repository = {
    statements: {
        patterns: ['#braces', '#comment', '#string'],
    },
    braces: {
        name: 'braces.s',
        patterns: ['#braces_paren', '#braces_curly'],
    },
    braces_paren: {
        captures: 'paren.braces.s',
        contentName: 'content.braces.s',
        begin: '(',
        end: ')',
        patterns: ['$self'],
    },
    braces_curly: {
        captures: 'curly.braces.s',
        contentName: 'content.braces.s',
        begin: '{',
        end: '}',
        patterns: ['$self'],
    },
    comment: {
        patterns: ['#comment_line', '#comment_doc_block', '#comment_block'],
    },
    comment_line: {
        patterns: ['#comment_line_double_slash', '#comment_line_hash'],
    },
    comment_line_hash: {
        name: 'comment.line.s',
        match: /(#).*\s?/,
        captures: {
            '1': 'punctuation.s',
        },
    },
    comment_line_double_slash: {
        name: 'comment.line.s',
        match: /(\/\/).*\s?/,
        captures: {
            '1': 'punctuation.s',
        },
    },
    comment_doc_block: {
        name: 'comment.block.s',
        begin: '/**',
        end: '*/',
        captures: {
            '0': 'punctuation.s',
        },
    },
    comment_block: {
        name: 'comment.block.s',
        begin: '/*',
        end: '*/',
        captures: {
            '0': 'punctuation.s',
        },
    },
    string: {
        patterns: [{ include: '#string_q_single' }, { include: '#string_q_double' }, { include: '#string_template' }],
    },
    string_q_single: {
        name: 'string.quoted.single.s',
        begin: "'",
        end: /'|((?:[^\\\n])$)/,
        patterns: [{ include: '#string_character_escape' }],
        captures: {
            '0': 'punctuation.s',
        },
    },
    string_q_double: {
        name: 'string.quoted.double.s',
        begin: '"',
        end: /"|((?:[^\\\n])$)/,
        patterns: [{ include: '#string_character_escape' }],
        captures: {
            '0': 'punctuation.s',
        },
    },
    string_template: {
        name: 'string.template.s',
        begin: '`',
        end: '`',
        patterns: [{ include: '#string_character_escape' }, '#template_embedded'],
        captures: {
            '0': 'punctuation.s',
        },
    },
    string_character_escape: {
        name: 'constant.character.escape.s',
        match: /\\(x[0-9A-Fa-f]{2}|[0-3][0-7]{0,2}|[4-7][0-7]?|.|$)/,
    },
    template_embedded: {
        name: 'embedded.template.s',
        begin: '${',
        end: '}',
        patterns: ['#statements'],
        captures: {
            '0': 'punctuation.s',
        },
    },
};

export const grammar: GrammarDef = {
    name: 'Simple',
    scopeName: 'source.s',
    patterns: ['#statements'],
    repository,
};
