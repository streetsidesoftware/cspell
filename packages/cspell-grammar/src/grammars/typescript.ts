import { Grammar, Repository } from '..';

const repository: Repository = {
    statements: {
        patterns: [{ include: '#string' }, { include: '#comment' }],
    },
    string: {
        patterns: [{ include: '#string_q_single' }, { include: '#string_q_double' }, { include: '#string_template' }],
    },
    string_q_single: {
        name: 'string.quoted.single.ts',
        begin: "'",
        end: /'|((?:[^\\\n])$)/,
        patterns: [{ include: '#string_character_escape' }],
    },
    string_q_double: {
        name: 'string.quoted.double.ts',
        begin: '"',
        end: /"|((?:[^\\\n])$)/,
        patterns: [{ include: '#string_character_escape' }],
    },
    string_template: {
        name: 'string.template.ts',
        begin: '`',
        end: '`',
        patterns: [{ include: '#string_character_escape' }],
    },
    string_wrap: {
        match: /(?:[^\\\n])$/,
    },
    string_character_escape: {
        name: 'constant.character.escape.ts',
        match: /\\(x[0-9A-Fa-f]{2}|[0-3][0-7]{0,2}|[4-7][0-7]?|.|$)/,
    },
    comment: {
        patterns: [{ include: '#comment_line' }, { include: '#comment_block' }],
    },
    comment_line: {
        name: 'comment.line.ts',
        match: /\/\/.*/,
    },
    comment_block: {
        name: 'comment.block.ts',
        begin: '/*',
        end: '*/',
    },
};

export const grammar: Grammar = {
    name: 'TypeScript',
    scopeName: 'source.ts',
    patterns: [
        {
            name: 'comment.line.shebang.ts',
            match: /^#!.*(?=$)/,
        },
        {
            include: '#statements',
        },
    ],
    repository,
};
