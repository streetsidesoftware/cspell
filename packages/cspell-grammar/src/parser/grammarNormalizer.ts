import type { GrammarDef, Pattern, Repository } from '..';
import type {
    Captures,
    PatternBeginEnd,
    PatternInclude,
    PatternList,
    PatternMatch,
    PatternName,
    PatternPatterns,
} from './grammarDefinition';
import type {
    GrammarRule,
    MatchRuleResult,
    NCaptures,
    NGrammar,
    NPattern,
    NPatternBeginEnd,
    NPatternInclude,
    NPatternMatch,
    NPatternName,
    NPatternPatterns,
    NPatternRepositoryReference,
    NRepository,
    NScope,
    NScopeSource,
    Rule,
} from './grammarNormalized';
import { isPatternBeginEnd, isPatternInclude, isPatternMatch, isPatternPatterns } from './grammarTypesHelpers';
import { createMatchResult, createSimpleMatchResult } from './matchResult';
import type { Scope } from './scope';
import { ScopePool } from './scope';
import type { LineOffsetAnchored, MatchResult } from './types';

export function normalizeGrammar(grammar: GrammarDef): NGrammar {
    return new ImplNGrammar(grammar);
}

const SpecialRepositoryReferences: Record<string, true | undefined> = {
    $self: true,
    $base: true,
};

export function nPattern(p: PatternMatch): NPatternMatch;
export function nPattern(p: PatternInclude): NPatternInclude;
export function nPattern(p: PatternBeginEnd): NPatternBeginEnd;
export function nPattern(p: PatternPatterns): NPatternPatterns;
export function nPattern(p: PatternName): NPatternName;
export function nPattern(p: Pattern): NPattern;
export function nPattern(p: Pattern): NPattern {
    if (isPatternMatch(p)) return normalizePatternMatch(p);
    if (isPatternBeginEnd(p)) return normalizePatternBeginEnd(p);
    if (isPatternInclude(p)) return normalizePatternInclude(p);
    if (isPatternPatterns(p)) return normalizePatternsPatterns(p);
    return normalizePatternName(p);
}

function normalizePatternMatch(p: PatternMatch): NPatternMatch {
    const regExec = makeTestMatchFn(p.match);
    const self: NPatternMatch = {
        ...p,
        captures: normalizeCapture(p.captures),
        findMatch,
    };

    function findMatch(line: LineOffsetAnchored, parentRule: Rule): MatchRuleResult | undefined {
        const match = regExec(line);
        if (!match) return undefined;
        const rule = factoryRule(parentRule, self);
        return { rule, match, line };
    }

    return self;
}

function normalizePatternBeginEnd(p: PatternBeginEnd): NPatternBeginEnd {
    const patterns = normalizePatterns(p.patterns);
    const self: NPatternBeginEnd = {
        ...p,
        captures: normalizeCapture(p.captures),
        beginCaptures: normalizeCapture(p.beginCaptures),
        endCaptures: normalizeCapture(p.endCaptures),
        patterns,
        findMatch,
    };

    function findMatch(line: LineOffsetAnchored, parentRule: Rule): MatchRuleResult | undefined {
        const match = testBegin(line);
        if (!match) return undefined;
        const rule = factoryRule(parentRule, self, findNext, end);
        return { rule, match, line };
    }

    const testBegin = makeTestMatchFn(p.begin);
    const testEnd = p.end !== undefined ? makeTestMatchFn(p.end) : () => undefined;

    function findNext(this: Rule, line: LineOffsetAnchored) {
        return patterns && findInPatterns(patterns, line, this);
    }

    function end(line: LineOffsetAnchored): MatchResult | undefined {
        return testEnd(line);
    }

    return self;
}

function normalizePatternName(p: PatternName): NPatternName {
    const patterns = undefined;
    const self: NPatternName = {
        ...p,
        patterns,
        findMatch,
    };

    function findMatch(line: LineOffsetAnchored, parentRule: Rule): MatchRuleResult | undefined {
        const rule = factoryRule(parentRule, self);
        const input = line.text.slice(line.offset);
        const match = createSimpleMatchResult(input, input, line.offset, line.lineNumber);
        return { rule, match, line };
    }

    return self;
}

function normalizePatternInclude(p: PatternInclude): NPatternInclude | NPatternRepositoryReference {
    const { include } = p;
    return include.startsWith('#') || include in SpecialRepositoryReferences
        ? normalizePatternIncludeRef(p)
        : normalizePatternIncludeExt(p);
}

function normalizePatternIncludeRef(p: PatternInclude): NPatternRepositoryReference {
    const { include, ...rest } = p;
    const reference = include.startsWith('#') ? include.slice(1) : include;
    const self: NPatternRepositoryReference = {
        ...rest,
        reference,
        findMatch,
    };

    function findMatch(line: LineOffsetAnchored, parentRule: Rule): MatchRuleResult | undefined {
        const pat = parentRule.repository[reference];
        if (pat === undefined) throw new Error(`Unknown Include Reference ${include}`);
        return pat.findMatch(line, parentRule);
    }

    return self;
}

function normalizePatternIncludeExt(p: PatternInclude): NPatternInclude | NPatternRepositoryReference {
    function findMatch(_line: LineOffsetAnchored): MatchRuleResult | undefined {
        return undefined;
    }

    const self: NPatternInclude = {
        ...p,
        findMatch,
    };

    return self;
}

function normalizePatternsPatterns(p: PatternPatterns): NPatternPatterns {
    return new ImplNPatternPatterns(p);
}

function findInPatterns(patterns: NPattern[], line: LineOffsetAnchored, rule: Rule): MatchRuleResult | undefined {
    let r: MatchRuleResult | undefined = undefined;
    for (const pat of patterns) {
        if (pat.disabled) continue;
        const er = pat.findMatch(line, rule);
        if (er?.match !== undefined && !er.rule.pattern.disabled) {
            r = (r && r.match && r.match.index <= er.match.index && r) || er;
        }
    }
    return r;
}

function normalizePatterns(patterns: undefined): undefined;
function normalizePatterns(patterns: PatternList): NPattern[];
function normalizePatterns(patterns: PatternList | undefined): NPattern[] | undefined;
function normalizePatterns(patterns: PatternList | undefined): NPattern[] | undefined {
    if (!patterns) return undefined;
    return patterns.map((p) => (typeof p === 'string' ? { include: p } : p)).map(nPattern);
}

const emptyRepository: NRepository = Object.freeze(Object.create(null));

function normalizePatternRepository(rep: Repository | undefined): NRepository {
    if (!rep) return emptyRepository;

    return normalizeRepository(rep);
}

function normalizeRepository(rep: Repository): NRepository {
    const repository: NRepository = Object.create(null);
    for (const [key, pat] of Object.entries(rep)) {
        repository[key] = nPattern(pat);
    }
    return repository;
}

let ruleCounter = 0;

function factoryRuleBase(
    parent: Rule | undefined,
    pattern: NGrammar,
    repository: NRepository,
    grammar: NGrammar,
    findNext: (this: Rule, line: LineOffsetAnchored) => MatchRuleResult | undefined,
    end: (this: Rule, line: LineOffsetAnchored) => MatchResult | undefined
): GrammarRule;
function factoryRuleBase(
    parent: Rule | undefined,
    pattern: NPattern | NGrammar,
    repository: NRepository,
    grammar: NGrammar,
    findNext?: (this: Rule, line: LineOffsetAnchored) => MatchRuleResult | undefined,
    end?: (this: Rule, line: LineOffsetAnchored) => MatchResult | undefined
): Rule;
function factoryRuleBase(
    parent: Rule | undefined,
    pattern: NPattern | NGrammar,
    repository: NRepository,
    grammar: NGrammar,
    findNext?: (this: Rule, line: LineOffsetAnchored) => MatchRuleResult | undefined,
    end?: (this: Rule, line: LineOffsetAnchored) => MatchResult | undefined
): Rule {
    const depth = parent ? parent.depth + 1 : 0;
    return {
        id: ruleCounter++,
        grammar,
        pattern,
        parent,
        repository,
        depth,
        findNext,
        end,
    };
}

function factoryRule(
    parent: Rule,
    pattern: NPattern,
    findNext?: (this: Rule, line: LineOffsetAnchored) => MatchRuleResult | undefined,
    end?: (this: Rule, line: LineOffsetAnchored) => MatchResult | undefined
): Rule {
    return factoryRuleBase(parent, pattern, parent.repository, parent.grammar, findNext, end);
}

function normalizeCapture(cap: Captures | undefined): NCaptures | undefined {
    if (cap === undefined) return undefined;
    if (typeof cap === 'string') return { [0]: cap };

    const capture: NCaptures = Object.create(null);
    for (const [key, pat] of Object.entries(cap)) {
        capture[key] = typeof pat === 'string' ? pat : normalizePatternName(pat).name;
    }

    return capture;
}

function makeTestMatchFn(reg: string | RegExp): (line: LineOffsetAnchored) => MatchResult | undefined {
    if (typeof reg === 'string') return matchString(reg);
    return matchRegExp(reg);
}

function matchString(s: string): (line: LineOffsetAnchored) => MatchResult | undefined {
    return (line) => {
        const input = line.text;
        const index = input.indexOf(s, line.offset);
        if (index < 0) return undefined;
        return createSimpleMatchResult(s, input, index, line.lineNumber);
    };
}

function matchRegExp(r: RegExp): (line: LineOffsetAnchored) => MatchResult | undefined {
    return (line) => {
        const rg = RegExp(r, 'gm');
        rg.lastIndex = line.offset;
        const m = rg.exec(line.text);
        return (m && createMatchResult(m, line.lineNumber)) ?? undefined;
    };
}

export function extractScope(er: Rule, isContent = true): Scope {
    const scope: string[] = [];

    for (let rule: Rule | undefined = er; rule; rule = rule.parent) {
        const pattern = rule.pattern;
        const { name, contentName } = pattern;
        if (contentName && isContent) {
            scope.push(contentName);
        }
        if (name !== undefined) {
            scope.push(name);
        }
        isContent = true;
    }

    return er.grammar.scopePool.parseScope(scope);
}

class ImplNGrammar implements NGrammar {
    readonly scopeName: NScopeSource;
    readonly name: NScopeSource;
    readonly comment: string | undefined;
    readonly disabled: boolean | undefined;
    readonly patterns: NPattern[];
    readonly repository: NRepository;
    readonly grammarName: string | undefined;
    readonly self: NPatternPatterns;
    readonly scopePool: ScopePool;

    constructor(grammar: GrammarDef) {
        this.scopeName = grammar.scopeName;
        this.name = grammar.scopeName;
        this.comment = grammar.comment;
        this.disabled = grammar.disabled;
        this.grammarName = grammar.name;

        const self = nPattern({
            patterns: [{ patterns: grammar.patterns }],
        });
        const repository = normalizePatternRepository(grammar.repository);

        this.patterns = self.patterns;
        this.repository = repository;
        this.self = self;
        this.scopePool = new ScopePool();
    }

    begin(parentRule?: Rule): GrammarRule {
        const patterns = this.patterns;

        function grammarToRule(grammar: NGrammar, baseGrammar: NGrammar, parent: Rule | undefined): GrammarRule {
            const repository: NRepository = Object.create(null);
            Object.assign(repository, grammar.repository);
            repository['$self'] = grammar.self;
            repository['$base'] = repository['$base'] || baseGrammar.self;

            function findNext(this: Rule, line: LineOffsetAnchored): MatchRuleResult | undefined {
                return findInPatterns(patterns, line, this);
            }
            function end(_line: LineOffsetAnchored) {
                return undefined;
            }

            return factoryRuleBase(parent, grammar, repository, grammar, findNext, end);
        }

        return grammarToRule(this, parentRule?.grammar ?? this, parentRule);
    }
}

class ImplNPatternPatterns implements NPatternPatterns {
    readonly name: NScope | undefined;
    readonly comment: string | undefined;
    readonly disabled: boolean | undefined;
    readonly patterns: NPattern[];

    constructor(p: PatternPatterns) {
        const { name, comment, disabled, ...rest } = p;
        this.patterns = normalizePatterns(rest.patterns);
        this.name = name;
        this.comment = comment;
        this.disabled = disabled;
    }

    findMatch(line: LineOffsetAnchored, parentRule: Rule): MatchRuleResult | undefined {
        const patterns = this.patterns;
        const rule = factoryRule(parentRule, this, findNext);
        function findNext(this: Rule, line: LineOffsetAnchored): MatchRuleResult | undefined {
            return findInPatterns(patterns, line, this);
        }
        return rule.findNext?.(line);
    }
}
