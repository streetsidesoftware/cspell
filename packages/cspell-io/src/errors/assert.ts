import { AssertionError } from './errors.js';

export function assert(value: unknown, message?: string): asserts value {
    if (!value) {
        throw new AssertionError(message ?? 'Assertion failed');
    }
}
