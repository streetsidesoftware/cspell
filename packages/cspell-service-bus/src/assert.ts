export function assert(value: unknown, message?: string | Error): asserts value {
    if (!value) {
        const err = message instanceof Error ? message : Error(message ?? 'AssertionError');
        throw err;
    }
}
