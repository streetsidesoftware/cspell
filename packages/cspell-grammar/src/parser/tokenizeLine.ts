import assert from 'assert';

import type { NGrammar, Rule } from './grammarNormalized.js';
import { extractScope } from './grammarNormalizer.js';
import { applyCaptureToBeginOrMatch, applyCaptureToEnd } from './processors/procMatchingRule.js';
import type { DocumentLine, LineOffsetAnchored, TokenizedLine, TokenizedLineResult, TokenizedText } from './types.js';

export function tokenizeLine(line: DocumentLine, rule: Rule): TokenizedLineResult {
    const text = line.text;
    const lineLen = line.text.length;
    const parsedText: TokenizedText[] = [];

    let ctx = buildContext({ ...line, offset: 0, anchor: -1 }, rule);

    // Note: the begin rule has already been matched and applied.
    // Does it need to end or go deeper?

    while (ctx.line.offset <= lineLen) {
        // We are at an end
        let endMatch = ctx.rule.end?.(ctx.line);
        while (endMatch?.index === ctx.line.offset) {
            parsedText.push(...applyCaptureToEnd(ctx.rule, endMatch));
            ctx = findParentWithEnd(ctx);
            ctx.line.offset = endMatch.index + endMatch.match.length;
            endMatch = ctx.rule.end?.(ctx.line);
        }

        if (ctx.line.offset >= lineLen) break;

        const { line, rule } = ctx;
        const offset = line.offset;
        const match = rule.findNext?.(line);
        const limit = endMatch?.index ?? lineLen;
        const emitTil = match ? Math.min(match.match.index, limit) : limit;
        if (offset < emitTil) {
            const scope = extractScope(rule);
            const start = offset;
            const end = emitTil;
            parsedText.push({
                scope,
                text: text.slice(start, end),
                range: [start, end, line.lineNumber],
            });
            ctx.line.offset = emitTil;
        }
        if (!match || (endMatch && endMatch.index <= match.match.index)) {
            continue; // We are at an end, process it at the beginning of the loop.
        }
        // Process the beginning of the match
        parsedText.push(...applyCaptureToBeginOrMatch(match));
        line.anchor = match.match.index + match.match.match.length;
        line.offset = line.anchor;
        ctx = findNearestWithEnd(buildContext(line, match.rule));
    }

    return toParseLineResult(line, ctx.rule, parsedText);
}

export function tokenizeText(text: string, grammar: NGrammar): TokenizedLine[] {
    return [...tokenizeTextIterable(text, grammar)];
}

export function* tokenizeTextIterable(text: string, grammar: NGrammar): Iterable<TokenizedLine> {
    const lines = text.split(/(?<=\r\n|\n|\r(?!\n))/);
    const rule = grammar.begin();
    let documentOffset = 0;
    let tr = tokenizeLine({ text: lines[0], lineNumber: 0, documentOffset }, rule);
    documentOffset += lines[0].length;
    yield toParsedLine(tr);
    for (let i = 1; i < lines.length; ++i) {
        const line = { text: lines[i], lineNumber: i, documentOffset };
        documentOffset += line.text.length;
        tr = tr.parse(line);
        yield toParsedLine(tr);
    }
}

function toParsedLine(pr: TokenizedLineResult): TokenizedLine {
    const { tokens: parsedText, line, offset } = pr;
    return { tokens: parsedText, line, offset };
}

function toParseLineResult(line: DocumentLine, rule: Rule, parsedText: TokenizedText[]): TokenizedLineResult {
    return {
        tokens: parsedText,
        line,
        offset: line.documentOffset,
        parse: (line: DocumentLine) => tokenizeLine(line, rule),
    };
}

interface Context {
    line: LineOffsetAnchored;
    rule: Rule;
    parent?: Context;
}

function buildContext(line: LineOffsetAnchored, rule: Rule): Context {
    const rules = calcRuleStack(rule);
    const numRules = rules.length;
    const rootNum = numRules - 1;
    const rootRule = rules[rootNum];

    const rootContext: Context = {
        line,
        rule: rootRule,
    };

    let ctx = rootContext;
    for (let i = rootNum - 1; i >= 0; --i) {
        const rule = rules[i];
        const line = ctx.line;
        ctx = {
            line,
            rule,
            parent: ctx,
        };
        // Check while here.
    }

    return ctx;
}

function calcRuleStack(rule: Rule) {
    const rules: Rule[] = [];
    let r: Rule | undefined = rule;
    while (r) {
        rules.push(r);
        r = r.parent;
    }
    return rules;
}

function must<T>(t: T | undefined, msg = 'Must be defined'): T {
    assert(t !== undefined && t !== null, msg);
    return t;
}

function findParentWithEnd(ctx: Context): Context {
    return findNearestWithEnd(must(ctx.parent));
}

function findNearestWithEnd(ctx: Context): Context {
    while (!ctx.rule.end) {
        ctx = must(ctx.parent);
    }
    return ctx;
}
