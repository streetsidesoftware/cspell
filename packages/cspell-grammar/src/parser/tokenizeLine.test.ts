import assert from 'assert';
import * as Simple from '../grammars/simple';
import { normalizeGrammar } from './grammarNormalizer';
import { tokenizeText } from './tokenizeLine';
import type { TokenizedLine } from './types';

// const oc = expect.objectContaining;

const grammar = normalizeGrammar(Simple.grammar);

describe('tokenizeLine', () => {
    interface TextAndName {
        text: string;
        name: string;
    }

    const sampleNestedParen = `
(
    {
        {
            deep
        }
    }
)
`;

    const sampleWithStrings = `
    const x = 'The auto\\'s left wheel.';
    function y(param = "first") {
        start = '('
        end = ")"
        msg = \`from \${start} to \${end} \`
    }
    `;

    function t(text: string, name?: string): TextAndName {
        return { text, name: name ?? JSON.stringify(text) };
    }

    test.each`
        test                                         | comment
        ${t('line # comment.')}                      | ${''}
        ${t('line # comment.\n')}                    | ${''}
        ${t('line ( a (b))')}                        | ${''}
        ${t('line ( a { b } c)')}                    | ${''}
        ${t(sampleNestedParen, 'sampleNestedParen')} | ${''}
        ${t(sampleWithStrings, 'complex sample')}    | ${''}
    `('tokenizeText $test.name - $comment', ({ test }: { test: TextAndName }) => {
        expect.addSnapshotSerializer({
            test: isTokenizedLine,
            serialize: serializeTokenizedLine,
        });
        const r = tokenizeText(test.text, grammar);
        assertParsedLinesAreValid(r);
        expect(r).toMatchSnapshot();
    });
});

// function ocScope(...scope: string[]) {
//     return oc({ scope });
// }

// function ocScopes(...scopes: string[][]) {
//     return scopes.map((s) => ocScope(...s));
// }

// function ocParsedTextScopes(...scopes: string[][]) {
//     return oc({
//         parsedText: ocScopes(...scopes),
//     });
// }

function assertParsedLinesAreValid(lines: TokenizedLine[]): asserts lines {
    lines.forEach(assertParsedLineIsValid);
}

function assertParsedLineIsValid(pLine: TokenizedLine): asserts pLine {
    const text = pLine.line.text;
    const joined = pLine.tokens.map((pt) => pt.text).join('');
    assert(text === joined);
    pLine.tokens.forEach((pt) => assert(text.startsWith(pt.text, pt.offset)));
}

/**
 * Serialize a TokenizedLine for snapshots to improve diffs.
 * One line for the line and one line each for each segment
 */
function serializeTokenizedLine(
    val: TokenizedLine,
    _config: unknown,
    indentation: string,
    _depth: number,
    _refs: unknown
) {
    const { line, tokens: parsedText } = val;
    const textAndScope = parsedText.map((t) => [
        `${t.offset}: ${JSON.stringify(t.text.replace(/\r/g, '↤').replace(/\n/g, '↩'))}`,
        `${t.scope.join(' ')}`,
    ]);

    const maxLen = textAndScope.reduce((a, ts) => Math.max(a, ts[0].length), 0);

    const pt = textAndScope
        .map((ts) => `${indentation}  ${ts[0]}${' '.repeat(maxLen - ts[0].length)}     -- ${ts[1]}`)
        .join('\n');
    return `${line.lineNumber}: ${JSON.stringify(line.text.replace(/\r/g, '↤').replace(/\n/g, '↩'))}:\n${pt}`;
}

function isTokenizedLine(v: unknown | TokenizedLine): v is TokenizedLine {
    if (!v || typeof v !== 'object') return false;
    const tl = <TokenizedLine>v;
    return typeof tl.line === 'object' && Array.isArray(tl.tokens);
}
