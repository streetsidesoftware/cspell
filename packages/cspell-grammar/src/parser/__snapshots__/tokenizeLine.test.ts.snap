// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizeLine tokenizeText "line # comment." -  1`] = `
Array [
  0: "line # comment.":
    0: "line "         -- source.s
    5: "#"             -- punctuation.s comment.line.s source.s
    6: " comment."     -- comment.line.s source.s,
]
`;

exports[`tokenizeLine tokenizeText "line # comment.\\n" -  1`] = `
Array [
  0: "line # comment.↩":
    0: "line "          -- source.s
    5: "#"              -- punctuation.s comment.line.s source.s
    6: " comment.↩"     -- comment.line.s source.s,
]
`;

exports[`tokenizeLine tokenizeText "line ( a (b))" -  1`] = `
Array [
  0: "line ( a (b))":
    0: "line "     -- source.s
    5: "("         -- paren.braces.s braces.s source.s
    6: " a "       -- content.braces.s braces.s source.s
    9: "("         -- paren.braces.s braces.s content.braces.s braces.s source.s
    10: "b"        -- content.braces.s braces.s content.braces.s braces.s source.s
    11: ")"        -- paren.braces.s braces.s content.braces.s braces.s source.s
    12: ")"        -- paren.braces.s braces.s source.s,
]
`;

exports[`tokenizeLine tokenizeText "line ( a { b } c)" -  1`] = `
Array [
  0: "line ( a { b } c)":
    0: "line "     -- source.s
    5: "("         -- paren.braces.s braces.s source.s
    6: " a "       -- content.braces.s braces.s source.s
    9: "{"         -- curly.braces.s braces.s content.braces.s braces.s source.s
    10: " b "      -- content.braces.s braces.s content.braces.s braces.s source.s
    13: "}"        -- curly.braces.s braces.s content.braces.s braces.s source.s
    14: " c"       -- content.braces.s braces.s source.s
    16: ")"        -- paren.braces.s braces.s source.s,
]
`;

exports[`tokenizeLine tokenizeText complex sample -  1`] = `
Array [
  0: "↩":
    0: "↩"     -- source.s,
  1: "    const x = 'The auto\\\\'s left wheel.';↩":
    0: "    const x = "     -- source.s
    14: "'"                 -- punctuation.s string.quoted.single.s source.s
    15: "The auto"          -- string.quoted.single.s source.s
    23: "\\\\'"               -- constant.character.escape.s string.quoted.single.s source.s
    25: "s left wheel."     -- string.quoted.single.s source.s
    38: "'"                 -- punctuation.s string.quoted.single.s source.s
    39: ";↩"                -- source.s,
  2: "    function y(param = \\"first\\") {↩":
    0: "    function y"     -- source.s
    14: "("                 -- paren.braces.s braces.s source.s
    15: "param = "          -- content.braces.s braces.s source.s
    23: "\\""                -- punctuation.s string.quoted.double.s content.braces.s braces.s source.s
    24: "first"             -- string.quoted.double.s content.braces.s braces.s source.s
    29: "\\""                -- punctuation.s string.quoted.double.s content.braces.s braces.s source.s
    30: ")"                 -- paren.braces.s braces.s source.s
    31: " "                 -- source.s
    32: "{"                 -- curly.braces.s braces.s source.s
    33: "↩"                 -- content.braces.s braces.s source.s,
  3: "        start = '('↩":
    0: "        start = "     -- content.braces.s braces.s source.s
    16: "'"                   -- punctuation.s string.quoted.single.s content.braces.s braces.s source.s
    17: "("                   -- string.quoted.single.s content.braces.s braces.s source.s
    18: "'"                   -- punctuation.s string.quoted.single.s content.braces.s braces.s source.s
    19: "↩"                   -- content.braces.s braces.s source.s,
  4: "        end = \\")\\"↩":
    0: "        end = "     -- content.braces.s braces.s source.s
    14: "\\""                -- punctuation.s string.quoted.double.s content.braces.s braces.s source.s
    15: ")"                 -- string.quoted.double.s content.braces.s braces.s source.s
    16: "\\""                -- punctuation.s string.quoted.double.s content.braces.s braces.s source.s
    17: "↩"                 -- content.braces.s braces.s source.s,
  5: "        msg = \`from \${start} to \${end} \`↩":
    0: "        msg = "     -- content.braces.s braces.s source.s
    14: "\`"                 -- punctuation.s string.template.s content.braces.s braces.s source.s
    15: "from "             -- string.template.s content.braces.s braces.s source.s
    20: "\${"                -- punctuation.s embedded.template.s string.template.s content.braces.s braces.s source.s
    22: "start"             -- embedded.template.s string.template.s content.braces.s braces.s source.s
    27: "}"                 -- punctuation.s embedded.template.s string.template.s content.braces.s braces.s source.s
    28: " to "              -- string.template.s content.braces.s braces.s source.s
    32: "\${"                -- punctuation.s embedded.template.s string.template.s content.braces.s braces.s source.s
    34: "end"               -- embedded.template.s string.template.s content.braces.s braces.s source.s
    37: "}"                 -- punctuation.s embedded.template.s string.template.s content.braces.s braces.s source.s
    38: " "                 -- string.template.s content.braces.s braces.s source.s
    39: "\`"                 -- punctuation.s string.template.s content.braces.s braces.s source.s
    40: "↩"                 -- content.braces.s braces.s source.s,
  6: "    }↩":
    0: "    "     -- content.braces.s braces.s source.s
    4: "}"        -- curly.braces.s braces.s source.s
    5: "↩"        -- source.s,
  7: "    ":
    0: "    "     -- source.s,
]
`;

exports[`tokenizeLine tokenizeText sampleNestedParen -  1`] = `
Array [
  0: "↩":
    0: "↩"     -- source.s,
  1: "(↩":
    0: "("     -- paren.braces.s braces.s source.s
    1: "↩"     -- content.braces.s braces.s source.s,
  2: "    {↩":
    0: "    "     -- content.braces.s braces.s source.s
    4: "{"        -- curly.braces.s braces.s content.braces.s braces.s source.s
    5: "↩"        -- content.braces.s braces.s content.braces.s braces.s source.s,
  3: "        {↩":
    0: "        "     -- content.braces.s braces.s content.braces.s braces.s source.s
    8: "{"            -- curly.braces.s braces.s content.braces.s braces.s content.braces.s braces.s source.s
    9: "↩"            -- content.braces.s braces.s content.braces.s braces.s content.braces.s braces.s source.s,
  4: "            deep↩":
    0: "            deep↩"     -- content.braces.s braces.s content.braces.s braces.s content.braces.s braces.s source.s,
  5: "        }↩":
    0: "        "     -- content.braces.s braces.s content.braces.s braces.s content.braces.s braces.s source.s
    8: "}"            -- curly.braces.s braces.s content.braces.s braces.s content.braces.s braces.s source.s
    9: "↩"            -- content.braces.s braces.s content.braces.s braces.s source.s,
  6: "    }↩":
    0: "    "     -- content.braces.s braces.s content.braces.s braces.s source.s
    4: "}"        -- curly.braces.s braces.s content.braces.s braces.s source.s
    5: "↩"        -- content.braces.s braces.s source.s,
  7: ")↩":
    0: ")"     -- paren.braces.s braces.s source.s
    1: "↩"     -- source.s,
]
`;
